File: mycodes/141-LinkedListCycle.java
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        ListNode p1 = head, p2 = head.next;
        while (p1 != p2) {
            if (p2 == null || p2.next == null) return false;
            p2 = p2.next.next;
            p1 = p1.next;
        }
        return true;
    }
}

File: mycodes/141-LinkedListCycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
  if head == nil || head.Next == nil {
    return false
  }
  s, f := head, head
  for f.Next != nil && f.Next.Next != nil {
    s = s.Next
    f = f.Next.Next
    if f == s {
      return true
    }
  }
  return false
}
